components/sync/driver/brave_sync_service_impl.h

```diff
@@ -54,21 +56,61 @@ class BraveSyncServiceImpl : public SyncServiceImpl {
 
   const brave_sync::Prefs& prefs() { return brave_sync_prefs_; }
 
+  void PermanentlyDeleteAccount(
+      base::OnceCallback<void(const SyncProtocolError&)> callback);
+
+  void SetJoinChainResultCallback(
+      base::OnceCallback<void(const bool&)> callback);
+
  private:
   friend class BraveSyncServiceImplTest;
   FRIEND_TEST_ALL_PREFIXES(BraveSyncServiceImplTest,
                            ForcedSetDecryptionPassphrase);
+  FRIEND_TEST_ALL_PREFIXES(BraveSyncServiceImplTest, OnAccountDeleted_Success);
+  FRIEND_TEST_ALL_PREFIXES(BraveSyncServiceImplTest,
+                           OnAccountDeleted_FailureAndRetry);
+  FRIEND_TEST_ALL_PREFIXES(BraveSyncServiceImplTest, JoinActiveOrNewChain);
+  FRIEND_TEST_ALL_PREFIXES(BraveSyncServiceImplTest, JoinDeletedChain);
+
   BraveSyncAuthManager* GetBraveSyncAuthManager();
   SyncServiceCrypto* GetCryptoForTests();
 
   void OnBraveSyncPrefsChanged(const std::string& path);
 
+  void PermanentlyDeleteAccountImpl(
+      const int current_attempt,
+      base::OnceCallback<void(const SyncProtocolError&)> callback);
+
+  void OnAccountDeleted(
+      const int current_attempt,
+      base::OnceCallback<void(const SyncProtocolError&)> callback,
+      const SyncProtocolError&);
+
+  void ResetEngine(ShutdownReason shutdown_reason,
+                   ResetEngineReason reset_reason) override;
+
+  void LocalDeviceAppeared();
+
   brave_sync::Prefs brave_sync_prefs_;
 
   PrefChangeRegistrar brave_sync_prefs_change_registrar_;
 
-  std::unique_ptr<SyncServiceImplDelegate> sync_service_impl_delegate_;
+  // This is set to true between |PermanentlyDeleteAccount| succeeded call and
+  // new sync chain setup or browser exit. This is used to avoid show the
+  // infobar to ourselves, because we know what we have done
+  bool initiated_delete_account_ = false;
 
+  // This flag is used to detect the case when we are trying to connect
+  // deleted sync chain. It is true between SetSyncCode and LocalDeviceAppeared.
+  bool initiated_join_chain_ = false;
```

Comment:
Why can't a new client join a deleted sync chain? If the old sync chain was already deleted, that means a new client can start over even if it is the same seed phrase.

Reply:
That's by design. We want to let users be able to fully delete a Sync chain and then block it from being used again (essentially that account in a *locked* state). The main use case is if you have lost one of your devices or or if it has been stolen. In that case, you don't want these devices re-uploading the data otherwise it won't be possible for you to ever delete your server-side account.

Reply:
That makes sense.